File: Data Structures & Algorithms/island-perimeter/submission-0.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    islandPerimeter(grid) {
        const ROWS = grid.length
        const COLS = grid[0].length

        const visited = Array.from({ length: ROWS }, () => 
            Array(COLS).fill(false)
        )
        console.log(visited)

        const dfs = (i, j) => {
            if (i < 0 || j < 0 || i >= ROWS || j >= COLS || grid[i][j] === 0) {
                return 1;
            }
            if (visited[i][j]) {
                return 0;
            }

            visited[i][j] = true;
            return (
                dfs(i, j + 1) + dfs(i + 1, j) + dfs(i, j - 1) + dfs(i - 1, j)
            );
        }

        for (let i = 0; i < ROWS; i++) {
            for (let j = 0; j < COLS; j++) {
                if (grid[i][j] === 1) {
                    return dfs(i, j)
                }
            }
        }
    }
}

File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.py
from typing import List

class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        rows, cols = len(heights), len(heights[0])

        pacific = set()
        atlantic = set()

        def dfs(r, c, visited, prev_height):
            if r < 0 or r >= rows or c < 0 or c >= cols: 
                return 

            if (r, c) in visited: 
                return

            
            if heights[r][c] < prev_height:
                return

            visited.add((r, c))

            dfs (r + 1, c, visited, heights[r][c])
            dfs (r - 1, c, visited, heights[r][c])
            dfs (r, c + 1, visited, heights[r][c])
            dfs (r, c - 1, visited, heights[r][c])

            
        for c in range(cols):
            dfs(0, c,  pacific, heights[0][c])

        for r in range(rows):
            dfs(r, 0,  pacific, heights[r][0])

        for c in range(cols):
            dfs(rows - 1, c, atlantic, heights[rows - 1][c])

        for r in range(rows):
            dfs(r, cols - 1, atlantic, heights[r][cols - 1])
        

        result = []

        for r in range(rows):
            for c in range(cols):
                if (r, c) in pacific and (r, c) in atlantic:
                    result.append([r, c])
        

        return result